packages/marketplace/contracts/Whitelist.sol

```diff
@@ -0,0 +1,138 @@
+    /// @param newErc20Permission allows to pay orders with only whitelisted token
+    /// @param whitelistsEnabled allows orders with any token */
+    // solhint-disable-next-line func-name-mixedcase
+    function __Whitelist_init(
+        address admin,
+        bytes32[] calldata roles,
+        bool[] calldata permissions,
+        bool whitelistsEnabled
+    ) internal onlyInitializing {
+        __AccessControlEnumerable_init_unchained();
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+        _setRolesEnabled(roles, permissions);
+        if (whitelistsEnabled) {
+            _enableWhitelists();
+        } else {
+            _disableWhitelists();
+        }
+    }
+
+    /// @notice setting permissions for tokens
+    /// @param roles we want to enable or disable
+    /// @param permissions boolan
+    function setRolesEnabled(
+        bytes32[] calldata roles,
+        bool[] calldata permissions
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _setRolesEnabled(roles, permissions);
+    }
+
+    /// @notice open market place for all non ERC20 tokens
+    function enableWhitelists() external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _enableWhitelists();
+    }
+
+    /// @notice setting permissions for open
+    function disableWhitelists() external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _disableWhitelists();
+    }
+
+    /// @notice Check if a specific role is enabled or disabled.
+    /// @param role The role identifier.
+    /// @return true if the role is enabled, false if disabled.
+    function isRoleEnabled(bytes32 role) public view returns (bool) {
+        return _rolesEnabled[role];
+    }
+
+    /// @notice Check if whitelists are enabled.
+    /// @return True if whitelists are enabled, false if disabled.
+    function isWhitelistsEnabled() public view returns (bool) {
+        return _whitelistsEnabled;
+    }
+
+    /// @notice setting permissions for tokens
+    /// @param roles identifyers
+    /// @param permissions booleans
+    function _setRolesEnabled(bytes32[] memory roles, bool[] memory permissions) internal {
+        for (uint256 i = 0; i < roles.length; ++i) {
+            if (_rolesEnabled[roles[i]] != permissions[i]) {
+                if (permissions[i]) {
+                    _enableRole(roles[i]);
+                } else {
+                    _disableRole(roles[i]);
+                }
+            }
+        }
+    }
+
+    function _enableRole(bytes32 role) internal {
+        _rolesEnabled[role] = true;
+        emit RoleEnabled(role);
+    }
+
+    function _disableRole(bytes32 role) internal {
+        _rolesEnabled[role] = false;
+        emit RoleDisabled(role);
+    }
+
+    function _enableWhitelists() internal {
+        _whitelistsEnabled = true;
+        emit WhitelistsEnabled();
+    }
+
+    function _disableWhitelists() internal {
+        _whitelistsEnabled = false;
+        emit WhitelistsDisabled();
+    }
+}
```

Comment:
aren't we missing `enableRole` and `disableRole` (external ones) ?

Reply:
that is done by setRolesEnabled